lib/domain.js

```diff
@@ -399,32 +399,32 @@ EventEmitter.init = function() {
 const eventEmit = EventEmitter.prototype.emit;
 EventEmitter.prototype.emit = function emit(...args) {
   const domain = this.domain;
-  if (domain === null || domain === undefined || this === process) {
-    return Reflect.apply(eventEmit, this, args);
-  }
 
   const type = args[0];
-  // edge case: if there is a domain and an existing non error object is given,
-  // it should not be errorized
-  // see test/parallel/test-event-emitter-no-error-provided-to-error-event.js
-  if (type === 'error' && args.length > 1 && args[1] &&
-    !(args[1] instanceof Error)) {
-    domain.emit('error', args[1]);
-    return false;
-  }
+  const doError = type === 'error' &&
+                  this.listenerCount(type) > 0;
 
-  domain.enter();
-  try {
+  if (doError || domain === null || domain === undefined || this === process) {
     return Reflect.apply(eventEmit, this, args);
-  } catch (er) {
+  }
+
+  if (type === 'error') {
+    const er = args.length > 1 && args[1] ?
```

Comment:
`args[1] || new errors.Error('ERR_UNHANDLED_ERROR')` should be enough. But I think it would be better to stay grammatically closer to the old implementation.

Something like:

```js
er = args[1]
if (!er) {
  const errors = lazyErrors();
  er = new errors.Error('ERR_UNHANDLED_ERROR');
}
if (typeof er === 'object' && er !== null) {
  er.domainEmitter = this;
  er.domain = domain;
  er.domainThrown = false;
}
```

Reply:
That triggers a slow path in V8. We need the `if` check or a ternary.